src/lsptoolshost/razor/htmlDocumentManager.ts

```diff
@@ -0,0 +1,118 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import { RazorLogger } from '../../razor/src/razorLogger';
+import { PlatformInformation } from '../../shared/platform';
+import { getUriPath } from '../../razor/src/uriPaths';
+import { virtualHtmlSuffix } from '../../razor/src/razorConventions';
+import { HtmlDocumentContentProvider } from './htmlDocumentContentProvider';
+import { HtmlDocument } from './htmlDocument';
+
+export class HtmlDocumentManager {
+    private readonly htmlDocuments: { [hostDocumentPath: string]: HtmlDocument } = {};
+    private readonly contentProvider: HtmlDocumentContentProvider;
+
+    constructor(private readonly platformInfo: PlatformInformation, private readonly logger: RazorLogger) {
+        this.contentProvider = new HtmlDocumentContentProvider(this, this.logger);
+    }
+
+    public get documents() {
+        return Object.values(this.htmlDocuments);
+    }
+
+    public register() {
+        const didCloseRegistration = vscode.workspace.onDidCloseTextDocument(async (document) => {
+            if (document.languageId !== 'html') {
+                return;
+            }
+
+            await this.closeDocument(document.uri);
+        });
+
+        const providerRegistration = vscode.workspace.registerTextDocumentContentProvider(
+            HtmlDocumentContentProvider.scheme,
+            this.contentProvider
+        );
+
+        return vscode.Disposable.from(didCloseRegistration, providerRegistration);
+    }
+
+    public async updateDocumentText(uri: vscode.Uri, text: string) {
+        const document = await this.getDocument(uri);
+
+        this.logger.logVerbose(`New content for '${uri}', updating '${document.path}'.`);
+
+        document.setContent(text);
+
+        this.contentProvider.fireDidChange(document.uri);
+    }
+
+    private async closeDocument(uri: vscode.Uri) {
+        const document = await this.getDocument(uri);
```

Comment:
Should this be `this.findDocument` so it doesn't add one if it didn't exist?

Reply:
Good catch, thanks

Reply:
Oh my god, thank you again. This comment totally found a bug, I was listening for the wrong document being closed, and if I didn't incorrectly call `getDocument` the whole thing would have stopped working. Fixed to listen to the Razor document now.